// allotment@1.20.3/dist/module.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import e, { useRef as t, useState as i, useEffect as n, useCallback as r, useMemo as s, useLayoutEffect as o, forwardRef as a, useImperativeHandle as u } from "react";
var c = "undefined" != typeof globalThis ? globalThis : "undefined" != typeof window ? window : "undefined" != typeof global ? global : "undefined" != typeof self ? self : {};
function h(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var l,
  f = {
    exports: {}
  };
/*!
  Copyright (c) 2018 Jed Watson.
  Licensed under the MIT License (MIT), see
  http://jedwatson.github.io/classnames
*/
var m,
  d,
  p,
  v = (l || (l = 1, m = f, function () {
    var e = {}.hasOwnProperty;
    function t() {
      for (var i = [], n = 0; n < arguments.length; n++) {
        var r = arguments[n];
        if (r) {
          var s = typeof r;
          if ("string" === s || "number" === s) i.push(r);else if (Array.isArray(r)) {
            if (r.length) {
              var o = t.apply(null, r);
              o && i.push(o);
            }
          } else if ("object" === s) if (r.toString === Object.prototype.toString) for (var a in r) e.call(r, a) && r[a] && i.push(a);else i.push(r.toString());
        }
      }
      return i.join(" ");
    }
    m.exports ? (t.default = t, m.exports = t) : window.classNames = t;
  }()), f.exports),
  y = h(v);
var b,
  g = h(function () {
    if (p) return d;
    p = 1;
    var e = NaN,
      t = "[object Symbol]",
      i = /^\s+|\s+$/g,
      n = /^[-+]0x[0-9a-f]+$/i,
      r = /^0b[01]+$/i,
      s = /^0o[0-7]+$/i,
      o = parseInt,
      a = Object.prototype.toString;
    function u(e) {
      var t = typeof e;
      return !!e && ("object" == t || "function" == t);
    }
    function c(c) {
      if ("number" == typeof c) return c;
      if (function (e) {
        return "symbol" == typeof e || function (e) {
          return !!e && "object" == typeof e;
        }(e) && a.call(e) == t;
      }(c)) return e;
      if (u(c)) {
        var h = "function" == typeof c.valueOf ? c.valueOf() : c;
        c = u(h) ? h + "" : h;
      }
      if ("string" != typeof c) return 0 === c ? c : +c;
      c = c.replace(i, "");
      var l = r.test(c);
      return l || s.test(c) ? o(c.slice(2), l ? 2 : 8) : n.test(c) ? e : +c;
    }
    return d = function d(e, t, i) {
      return void 0 === i && (i = t, t = void 0), void 0 !== i && (i = (i = c(i)) == i ? i : 0), void 0 !== t && (t = (t = c(t)) == t ? t : 0), function (e, t, i) {
        return e == e && (void 0 !== i && (e = e <= i ? e : i), void 0 !== t && (e = e >= t ? e : t)), e;
      }(c(e), t, i);
    };
  }()),
  z = {
    exports: {}
  };
var w = (b || (b = 1, function (e, t) {
    var i = "__lodash_hash_undefined__",
      n = 1,
      r = 2,
      s = 9007199254740991,
      o = "[object Arguments]",
      a = "[object Array]",
      u = "[object AsyncFunction]",
      h = "[object Boolean]",
      l = "[object Date]",
      f = "[object Error]",
      m = "[object Function]",
      d = "[object GeneratorFunction]",
      p = "[object Map]",
      v = "[object Number]",
      y = "[object Null]",
      b = "[object Object]",
      g = "[object Promise]",
      z = "[object Proxy]",
      w = "[object RegExp]",
      S = "[object Set]",
      _ = "[object String]",
      I = "[object Symbol]",
      x = "[object Undefined]",
      E = "[object WeakMap]",
      V = "[object ArrayBuffer]",
      j = "[object DataView]",
      L = /^\[object .+?Constructor\]$/,
      O = /^(?:0|[1-9]\d*)$/,
      D = {};
    D["[object Float32Array]"] = D["[object Float64Array]"] = D["[object Int8Array]"] = D["[object Int16Array]"] = D["[object Int32Array]"] = D["[object Uint8Array]"] = D["[object Uint8ClampedArray]"] = D["[object Uint16Array]"] = D["[object Uint32Array]"] = !0, D[o] = D[a] = D[V] = D[h] = D[j] = D[l] = D[f] = D[m] = D[p] = D[v] = D[b] = D[w] = D[S] = D[_] = D[E] = !1;
    var N = "object" == typeof c && c && c.Object === Object && c,
      M = "object" == typeof self && self && self.Object === Object && self,
      P = N || M || Function("return this")(),
      A = t && !t.nodeType && t,
      T = A && e && !e.nodeType && e,
      C = T && T.exports === A,
      F = C && N.process,
      k = function () {
        try {
          return F && F.binding && F.binding("util");
        } catch (e) {}
      }(),
      $ = k && k.isTypedArray;
    function H(e, t) {
      for (var i = -1, n = null == e ? 0 : e.length; ++i < n;) if (t(e[i], i, e)) return !0;
      return !1;
    }
    function Y(e) {
      var t = -1,
        i = Array(e.size);
      return e.forEach(function (e, n) {
        i[++t] = [n, e];
      }), i;
    }
    function B(e) {
      var t = -1,
        i = Array(e.size);
      return e.forEach(function (e) {
        i[++t] = e;
      }), i;
    }
    var R,
      W,
      G,
      U = Array.prototype,
      X = Function.prototype,
      J = Object.prototype,
      K = P["__core-js_shared__"],
      Z = X.toString,
      Q = J.hasOwnProperty,
      q = (R = /[^.]+$/.exec(K && K.keys && K.keys.IE_PROTO || "")) ? "Symbol(src)_1." + R : "",
      ee = J.toString,
      te = RegExp("^" + Z.call(Q).replace(/[\\^$.*+?()[\]{}|]/g, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"),
      ie = C ? P.Buffer : void 0,
      ne = P.Symbol,
      re = P.Uint8Array,
      se = J.propertyIsEnumerable,
      oe = U.splice,
      ae = ne ? ne.toStringTag : void 0,
      ue = Object.getOwnPropertySymbols,
      ce = ie ? ie.isBuffer : void 0,
      he = (W = Object.keys, G = Object, function (e) {
        return W(G(e));
      }),
      le = ke(P, "DataView"),
      fe = ke(P, "Map"),
      me = ke(P, "Promise"),
      de = ke(P, "Set"),
      pe = ke(P, "WeakMap"),
      ve = ke(Object, "create"),
      ye = Be(le),
      be = Be(fe),
      ge = Be(me),
      ze = Be(de),
      we = Be(pe),
      Se = ne ? ne.prototype : void 0,
      _e = Se ? Se.valueOf : void 0;
    function Ie(e) {
      var t = -1,
        i = null == e ? 0 : e.length;
      for (this.clear(); ++t < i;) {
        var n = e[t];
        this.set(n[0], n[1]);
      }
    }
    function xe(e) {
      var t = -1,
        i = null == e ? 0 : e.length;
      for (this.clear(); ++t < i;) {
        var n = e[t];
        this.set(n[0], n[1]);
      }
    }
    function Ee(e) {
      var t = -1,
        i = null == e ? 0 : e.length;
      for (this.clear(); ++t < i;) {
        var n = e[t];
        this.set(n[0], n[1]);
      }
    }
    function Ve(e) {
      var t = -1,
        i = null == e ? 0 : e.length;
      for (this.__data__ = new Ee(); ++t < i;) this.add(e[t]);
    }
    function je(e) {
      var t = this.__data__ = new xe(e);
      this.size = t.size;
    }
    function Le(e, t) {
      var i = Ge(e),
        n = !i && We(e),
        r = !i && !n && Ue(e),
        s = !i && !n && !r && Qe(e),
        o = i || n || r || s,
        a = o ? function (e, t) {
          for (var i = -1, n = Array(e); ++i < e;) n[i] = t(i);
          return n;
        }(e.length, String) : [],
        u = a.length;
      for (var c in e) !Q.call(e, c) || o && ("length" == c || r && ("offset" == c || "parent" == c) || s && ("buffer" == c || "byteLength" == c || "byteOffset" == c) || Ye(c, u)) || a.push(c);
      return a;
    }
    function Oe(e, t) {
      for (var i = e.length; i--;) if (Re(e[i][0], t)) return i;
      return -1;
    }
    function De(e) {
      return null == e ? void 0 === e ? x : y : ae && ae in Object(e) ? function (e) {
        var t = Q.call(e, ae),
          i = e[ae];
        try {
          e[ae] = void 0;
          var n = !0;
        } catch (e) {}
        var r = ee.call(e);
        return n && (t ? e[ae] = i : delete e[ae]), r;
      }(e) : function (e) {
        return ee.call(e);
      }(e);
    }
    function Ne(e) {
      return Ze(e) && De(e) == o;
    }
    function Me(e, t, i, s, u) {
      return e === t || (null == e || null == t || !Ze(e) && !Ze(t) ? e != e && t != t : function (e, t, i, s, u, c) {
        var m = Ge(e),
          d = Ge(t),
          y = m ? a : He(e),
          g = d ? a : He(t),
          z = (y = y == o ? b : y) == b,
          x = (g = g == o ? b : g) == b,
          E = y == g;
        if (E && Ue(e)) {
          if (!Ue(t)) return !1;
          m = !0, z = !1;
        }
        if (E && !z) return c || (c = new je()), m || Qe(e) ? Te(e, t, i, s, u, c) : function (e, t, i, s, o, a, u) {
          switch (i) {
            case j:
              if (e.byteLength != t.byteLength || e.byteOffset != t.byteOffset) return !1;
              e = e.buffer, t = t.buffer;
            case V:
              return !(e.byteLength != t.byteLength || !a(new re(e), new re(t)));
            case h:
            case l:
            case v:
              return Re(+e, +t);
            case f:
              return e.name == t.name && e.message == t.message;
            case w:
            case _:
              return e == t + "";
            case p:
              var c = Y;
            case S:
              var m = s & n;
              if (c || (c = B), e.size != t.size && !m) return !1;
              var d = u.get(e);
              if (d) return d == t;
              s |= r, u.set(e, t);
              var y = Te(c(e), c(t), s, o, a, u);
              return u.delete(e), y;
            case I:
              if (_e) return _e.call(e) == _e.call(t);
          }
          return !1;
        }(e, t, y, i, s, u, c);
        if (!(i & n)) {
          var L = z && Q.call(e, "__wrapped__"),
            O = x && Q.call(t, "__wrapped__");
          if (L || O) {
            var D = L ? e.value() : e,
              N = O ? t.value() : t;
            return c || (c = new je()), u(D, N, i, s, c);
          }
        }
        return !!E && (c || (c = new je()), function (e, t, i, r, s, o) {
          var a = i & n,
            u = Ce(e),
            c = u.length,
            h = Ce(t),
            l = h.length;
          if (c != l && !a) return !1;
          for (var f = c; f--;) {
            var m = u[f];
            if (!(a ? m in t : Q.call(t, m))) return !1;
          }
          var d = o.get(e);
          if (d && o.get(t)) return d == t;
          var p = !0;
          o.set(e, t), o.set(t, e);
          for (var v = a; ++f < c;) {
            var y = e[m = u[f]],
              b = t[m];
            if (r) var g = a ? r(b, y, m, t, e, o) : r(y, b, m, e, t, o);
            if (!(void 0 === g ? y === b || s(y, b, i, r, o) : g)) {
              p = !1;
              break;
            }
            v || (v = "constructor" == m);
          }
          if (p && !v) {
            var z = e.constructor,
              w = t.constructor;
            z == w || !("constructor" in e) || !("constructor" in t) || "function" == typeof z && z instanceof z && "function" == typeof w && w instanceof w || (p = !1);
          }
          return o.delete(e), o.delete(t), p;
        }(e, t, i, s, u, c));
      }(e, t, i, s, Me, u));
    }
    function Pe(e) {
      return !(!Ke(e) || function (e) {
        return !!q && q in e;
      }(e)) && (Xe(e) ? te : L).test(Be(e));
    }
    function Ae(e) {
      if (i = (t = e) && t.constructor, n = "function" == typeof i && i.prototype || J, t !== n) return he(e);
      var t,
        i,
        n,
        r = [];
      for (var s in Object(e)) Q.call(e, s) && "constructor" != s && r.push(s);
      return r;
    }
    function Te(e, t, i, s, o, a) {
      var u = i & n,
        c = e.length,
        h = t.length;
      if (c != h && !(u && h > c)) return !1;
      var l = a.get(e);
      if (l && a.get(t)) return l == t;
      var f = -1,
        m = !0,
        d = i & r ? new Ve() : void 0;
      for (a.set(e, t), a.set(t, e); ++f < c;) {
        var p = e[f],
          v = t[f];
        if (s) var y = u ? s(v, p, f, t, e, a) : s(p, v, f, e, t, a);
        if (void 0 !== y) {
          if (y) continue;
          m = !1;
          break;
        }
        if (d) {
          if (!H(t, function (e, t) {
            if (n = t, !d.has(n) && (p === e || o(p, e, i, s, a))) return d.push(t);
            var n;
          })) {
            m = !1;
            break;
          }
        } else if (p !== v && !o(p, v, i, s, a)) {
          m = !1;
          break;
        }
      }
      return a.delete(e), a.delete(t), m;
    }
    function Ce(e) {
      return function (e, t, i) {
        var n = t(e);
        return Ge(e) ? n : function (e, t) {
          for (var i = -1, n = t.length, r = e.length; ++i < n;) e[r + i] = t[i];
          return e;
        }(n, i(e));
      }(e, qe, $e);
    }
    function Fe(e, t) {
      var i,
        n,
        r = e.__data__;
      return ("string" == (n = typeof (i = t)) || "number" == n || "symbol" == n || "boolean" == n ? "__proto__" !== i : null === i) ? r["string" == typeof t ? "string" : "hash"] : r.map;
    }
    function ke(e, t) {
      var i = function (e, t) {
        return null == e ? void 0 : e[t];
      }(e, t);
      return Pe(i) ? i : void 0;
    }
    Ie.prototype.clear = function () {
      this.__data__ = ve ? ve(null) : {}, this.size = 0;
    }, Ie.prototype.delete = function (e) {
      var t = this.has(e) && delete this.__data__[e];
      return this.size -= t ? 1 : 0, t;
    }, Ie.prototype.get = function (e) {
      var t = this.__data__;
      if (ve) {
        var n = t[e];
        return n === i ? void 0 : n;
      }
      return Q.call(t, e) ? t[e] : void 0;
    }, Ie.prototype.has = function (e) {
      var t = this.__data__;
      return ve ? void 0 !== t[e] : Q.call(t, e);
    }, Ie.prototype.set = function (e, t) {
      var n = this.__data__;
      return this.size += this.has(e) ? 0 : 1, n[e] = ve && void 0 === t ? i : t, this;
    }, xe.prototype.clear = function () {
      this.__data__ = [], this.size = 0;
    }, xe.prototype.delete = function (e) {
      var t = this.__data__,
        i = Oe(t, e);
      return !(i < 0 || (i == t.length - 1 ? t.pop() : oe.call(t, i, 1), --this.size, 0));
    }, xe.prototype.get = function (e) {
      var t = this.__data__,
        i = Oe(t, e);
      return i < 0 ? void 0 : t[i][1];
    }, xe.prototype.has = function (e) {
      return Oe(this.__data__, e) > -1;
    }, xe.prototype.set = function (e, t) {
      var i = this.__data__,
        n = Oe(i, e);
      return n < 0 ? (++this.size, i.push([e, t])) : i[n][1] = t, this;
    }, Ee.prototype.clear = function () {
      this.size = 0, this.__data__ = {
        hash: new Ie(),
        map: new (fe || xe)(),
        string: new Ie()
      };
    }, Ee.prototype.delete = function (e) {
      var t = Fe(this, e).delete(e);
      return this.size -= t ? 1 : 0, t;
    }, Ee.prototype.get = function (e) {
      return Fe(this, e).get(e);
    }, Ee.prototype.has = function (e) {
      return Fe(this, e).has(e);
    }, Ee.prototype.set = function (e, t) {
      var i = Fe(this, e),
        n = i.size;
      return i.set(e, t), this.size += i.size == n ? 0 : 1, this;
    }, Ve.prototype.add = Ve.prototype.push = function (e) {
      return this.__data__.set(e, i), this;
    }, Ve.prototype.has = function (e) {
      return this.__data__.has(e);
    }, je.prototype.clear = function () {
      this.__data__ = new xe(), this.size = 0;
    }, je.prototype.delete = function (e) {
      var t = this.__data__,
        i = t.delete(e);
      return this.size = t.size, i;
    }, je.prototype.get = function (e) {
      return this.__data__.get(e);
    }, je.prototype.has = function (e) {
      return this.__data__.has(e);
    }, je.prototype.set = function (e, t) {
      var i = this.__data__;
      if (i instanceof xe) {
        var n = i.__data__;
        if (!fe || n.length < 199) return n.push([e, t]), this.size = ++i.size, this;
        i = this.__data__ = new Ee(n);
      }
      return i.set(e, t), this.size = i.size, this;
    };
    var $e = ue ? function (e) {
        return null == e ? [] : (e = Object(e), function (e, t) {
          for (var i = -1, n = null == e ? 0 : e.length, r = 0, s = []; ++i < n;) {
            var o = e[i];
            t(o, i, e) && (s[r++] = o);
          }
          return s;
        }(ue(e), function (t) {
          return se.call(e, t);
        }));
      } : function () {
        return [];
      },
      He = De;
    function Ye(e, t) {
      return !!(t = null == t ? s : t) && ("number" == typeof e || O.test(e)) && e > -1 && e % 1 == 0 && e < t;
    }
    function Be(e) {
      if (null != e) {
        try {
          return Z.call(e);
        } catch (e) {}
        try {
          return e + "";
        } catch (e) {}
      }
      return "";
    }
    function Re(e, t) {
      return e === t || e != e && t != t;
    }
    (le && He(new le(new ArrayBuffer(1))) != j || fe && He(new fe()) != p || me && He(me.resolve()) != g || de && He(new de()) != S || pe && He(new pe()) != E) && (He = function He(e) {
      var t = De(e),
        i = t == b ? e.constructor : void 0,
        n = i ? Be(i) : "";
      if (n) switch (n) {
        case ye:
          return j;
        case be:
          return p;
        case ge:
          return g;
        case ze:
          return S;
        case we:
          return E;
      }
      return t;
    });
    var We = Ne(function () {
        return arguments;
      }()) ? Ne : function (e) {
        return Ze(e) && Q.call(e, "callee") && !se.call(e, "callee");
      },
      Ge = Array.isArray,
      Ue = ce || function () {
        return !1;
      };
    function Xe(e) {
      if (!Ke(e)) return !1;
      var t = De(e);
      return t == m || t == d || t == u || t == z;
    }
    function Je(e) {
      return "number" == typeof e && e > -1 && e % 1 == 0 && e <= s;
    }
    function Ke(e) {
      var t = typeof e;
      return null != e && ("object" == t || "function" == t);
    }
    function Ze(e) {
      return null != e && "object" == typeof e;
    }
    var Qe = $ ? function (e) {
      return function (t) {
        return e(t);
      };
    }($) : function (e) {
      return Ze(e) && Je(e.length) && !!D[De(e)];
    };
    function qe(e) {
      return null != (t = e) && Je(t.length) && !Xe(t) ? Le(e) : Ae(e);
      var t;
    }
    e.exports = function (e, t) {
      return Me(e, t);
    };
  }(z, z.exports)), z.exports),
  S = h(w);
function _(e, t, i) {
  return e[t] ? e[t][0] ? e[t][0][i] : e[t][i] : "contentBoxSize" === t ? e.contentRect["inlineSize" === i ? "width" : "height"] : void 0;
}
function I(e) {
  void 0 === e && (e = {});
  var o = e.onResize,
    a = t(void 0);
  a.current = o;
  var u = e.round || Math.round,
    c = t(),
    h = i({
      width: void 0,
      height: void 0
    }),
    l = h[0],
    f = h[1],
    m = t(!1);
  n(function () {
    return m.current = !1, function () {
      m.current = !0;
    };
  }, []);
  var d = t({
      width: void 0,
      height: void 0
    }),
    p = function (e, i) {
      var s = t(null),
        o = t(null);
      o.current = i;
      var a = t(null);
      n(function () {
        u();
      });
      var u = r(function () {
        var t = a.current,
          i = o.current,
          n = t || (i ? i instanceof Element ? i : i.current : null);
        s.current && s.current.element === n && s.current.subscriber === e || (s.current && s.current.cleanup && s.current.cleanup(), s.current = {
          element: n,
          subscriber: e,
          cleanup: n ? e(n) : void 0
        });
      }, [e]);
      return n(function () {
        return function () {
          s.current && s.current.cleanup && (s.current.cleanup(), s.current = null);
        };
      }, []), r(function (e) {
        a.current = e, u();
      }, [u]);
    }(r(function (t) {
      return c.current && c.current.box === e.box && c.current.round === u || (c.current = {
        box: e.box,
        round: u,
        instance: new ResizeObserver(function (t) {
          var i = t[0],
            n = "border-box" === e.box ? "borderBoxSize" : "device-pixel-content-box" === e.box ? "devicePixelContentBoxSize" : "contentBoxSize",
            r = _(i, n, "inlineSize"),
            s = _(i, n, "blockSize"),
            o = r ? u(r) : void 0,
            c = s ? u(s) : void 0;
          if (d.current.width !== o || d.current.height !== c) {
            var h = {
              width: o,
              height: c
            };
            d.current.width = o, d.current.height = c, a.current ? a.current(h) : m.current || f(h);
          }
        })
      }), c.current.instance.observe(t, {
        box: e.box
      }), function () {
        c.current && c.current.instance.unobserve(t);
      };
    }, [e.box, u]), e.ref);
  return s(function () {
    return {
      ref: p,
      width: l.width,
      height: l.height
    };
  }, [p, l.width, l.height]);
}
var x = "allotment-module_splitView__L-yRc",
  E = "allotment-module_sashContainer__fzwJF",
  V = "allotment-module_splitViewContainer__rQnVa",
  j = "allotment-module_splitViewView__MGZ6O",
  L = "allotment-module_vertical__WSwwa",
  O = "allotment-module_horizontal__7doS8",
  D = "allotment-module_separatorBorder__x-rDS";
var N,
  M = !1,
  P = !1;
"object" == typeof navigator && (N = navigator.userAgent, P = N.indexOf("Macintosh") >= 0, M = (N.indexOf("Macintosh") >= 0 || N.indexOf("iPad") >= 0 || N.indexOf("iPhone") >= 0) && !!navigator.maxTouchPoints && navigator.maxTouchPoints > 0);
var A = M,
  T = P,
  C = "undefined" != typeof window && void 0 !== window.document && void 0 !== window.document.createElement ? o : n;
class F {
  constructor() {
    this._size = void 0;
  }
  getSize() {
    return this._size;
  }
  setSize(e) {
    this._size = e;
  }
}
function k(e, t) {
  var i = e.length,
    n = i - t.length;
  return n >= 0 && e.slice(n, i) === t;
}
var $,
  H = {
    exports: {}
  };
var Y,
  B,
  R = ($ || ($ = 1, function (e) {
    var t = Object.prototype.hasOwnProperty,
      i = "~";
    function n() {}
    function r(e, t, i) {
      this.fn = e, this.context = t, this.once = i || !1;
    }
    function s(e, t, n, s, o) {
      if ("function" != typeof n) throw new TypeError("The listener must be a function");
      var a = new r(n, s || e, o),
        u = i ? i + t : t;
      return e._events[u] ? e._events[u].fn ? e._events[u] = [e._events[u], a] : e._events[u].push(a) : (e._events[u] = a, e._eventsCount++), e;
    }
    function o(e, t) {
      0 == --e._eventsCount ? e._events = new n() : delete e._events[t];
    }
    function a() {
      this._events = new n(), this._eventsCount = 0;
    }
    Object.create && (n.prototype = Object.create(null), new n().__proto__ || (i = !1)), a.prototype.eventNames = function () {
      var e,
        n,
        r = [];
      if (0 === this._eventsCount) return r;
      for (n in e = this._events) t.call(e, n) && r.push(i ? n.slice(1) : n);
      return Object.getOwnPropertySymbols ? r.concat(Object.getOwnPropertySymbols(e)) : r;
    }, a.prototype.listeners = function (e) {
      var t = i ? i + e : e,
        n = this._events[t];
      if (!n) return [];
      if (n.fn) return [n.fn];
      for (var r = 0, s = n.length, o = new Array(s); r < s; r++) o[r] = n[r].fn;
      return o;
    }, a.prototype.listenerCount = function (e) {
      var t = i ? i + e : e,
        n = this._events[t];
      return n ? n.fn ? 1 : n.length : 0;
    }, a.prototype.emit = function (e, t, n, r, s, o) {
      var a = i ? i + e : e;
      if (!this._events[a]) return !1;
      var u,
        c,
        h = this._events[a],
        l = arguments.length;
      if (h.fn) {
        switch (h.once && this.removeListener(e, h.fn, void 0, !0), l) {
          case 1:
            return h.fn.call(h.context), !0;
          case 2:
            return h.fn.call(h.context, t), !0;
          case 3:
            return h.fn.call(h.context, t, n), !0;
          case 4:
            return h.fn.call(h.context, t, n, r), !0;
          case 5:
            return h.fn.call(h.context, t, n, r, s), !0;
          case 6:
            return h.fn.call(h.context, t, n, r, s, o), !0;
        }
        for (c = 1, u = new Array(l - 1); c < l; c++) u[c - 1] = arguments[c];
        h.fn.apply(h.context, u);
      } else {
        var f,
          m = h.length;
        for (c = 0; c < m; c++) switch (h[c].once && this.removeListener(e, h[c].fn, void 0, !0), l) {
          case 1:
            h[c].fn.call(h[c].context);
            break;
          case 2:
            h[c].fn.call(h[c].context, t);
            break;
          case 3:
            h[c].fn.call(h[c].context, t, n);
            break;
          case 4:
            h[c].fn.call(h[c].context, t, n, r);
            break;
          default:
            if (!u) for (f = 1, u = new Array(l - 1); f < l; f++) u[f - 1] = arguments[f];
            h[c].fn.apply(h[c].context, u);
        }
      }
      return !0;
    }, a.prototype.on = function (e, t, i) {
      return s(this, e, t, i, !1);
    }, a.prototype.once = function (e, t, i) {
      return s(this, e, t, i, !0);
    }, a.prototype.removeListener = function (e, t, n, r) {
      var s = i ? i + e : e;
      if (!this._events[s]) return this;
      if (!t) return o(this, s), this;
      var a = this._events[s];
      if (a.fn) a.fn !== t || r && !a.once || n && a.context !== n || o(this, s);else {
        for (var u = 0, c = [], h = a.length; u < h; u++) (a[u].fn !== t || r && !a[u].once || n && a[u].context !== n) && c.push(a[u]);
        c.length ? this._events[s] = 1 === c.length ? c[0] : c : o(this, s);
      }
      return this;
    }, a.prototype.removeAllListeners = function (e) {
      var t;
      return e ? (t = i ? i + e : e, this._events[t] && o(this, t)) : (this._events = new n(), this._eventsCount = 0), this;
    }, a.prototype.off = a.prototype.removeListener, a.prototype.addListener = a.prototype.on, a.prefixed = i, a.EventEmitter = a, e.exports = a;
  }(H)), H.exports),
  W = h(R);
function G(e, t) {
  var i = e.indexOf(t);
  i > -1 && (e.splice(i, 1), e.unshift(t));
}
function U(e, t) {
  var i = e.indexOf(t);
  i > -1 && (e.splice(i, 1), e.push(t));
}
function X(e, t, i) {
  if (i === void 0) {
    i = 1;
  }
  var n = Math.max(0, Math.ceil((t - e) / i)),
    r = new Array(n);
  var s = -1;
  for (; ++s < n;) r[s] = e + s * i;
  return r;
}
var J = h(function () {
    if (B) return Y;
    B = 1;
    var e = NaN,
      t = "[object Symbol]",
      i = /^\s+|\s+$/g,
      n = /^[-+]0x[0-9a-f]+$/i,
      r = /^0b[01]+$/i,
      s = /^0o[0-7]+$/i,
      o = parseInt,
      a = "object" == typeof c && c && c.Object === Object && c,
      u = "object" == typeof self && self && self.Object === Object && self,
      h = a || u || Function("return this")(),
      l = Object.prototype.toString,
      f = Math.max,
      m = Math.min,
      d = function d() {
        return h.Date.now();
      };
    function p(e) {
      var t = typeof e;
      return !!e && ("object" == t || "function" == t);
    }
    function v(a) {
      if ("number" == typeof a) return a;
      if (function (e) {
        return "symbol" == typeof e || function (e) {
          return !!e && "object" == typeof e;
        }(e) && l.call(e) == t;
      }(a)) return e;
      if (p(a)) {
        var u = "function" == typeof a.valueOf ? a.valueOf() : a;
        a = p(u) ? u + "" : u;
      }
      if ("string" != typeof a) return 0 === a ? a : +a;
      a = a.replace(i, "");
      var c = r.test(a);
      return c || s.test(a) ? o(a.slice(2), c ? 2 : 8) : n.test(a) ? e : +a;
    }
    return Y = function Y(e, t, i) {
      var n,
        r,
        s,
        o,
        a,
        u,
        c = 0,
        h = !1,
        l = !1,
        y = !0;
      if ("function" != typeof e) throw new TypeError("Expected a function");
      function b(t) {
        var i = n,
          s = r;
        return n = r = void 0, c = t, o = e.apply(s, i);
      }
      function g(e) {
        var i = e - u;
        return void 0 === u || i >= t || i < 0 || l && e - c >= s;
      }
      function z() {
        var e = d();
        if (g(e)) return w(e);
        a = setTimeout(z, function (e) {
          var i = t - (e - u);
          return l ? m(i, s - (e - c)) : i;
        }(e));
      }
      function w(e) {
        return a = void 0, y && n ? b(e) : (n = r = void 0, o);
      }
      function S() {
        var e = d(),
          i = g(e);
        if (n = arguments, r = this, u = e, i) {
          if (void 0 === a) return function (e) {
            return c = e, a = setTimeout(z, t), h ? b(e) : o;
          }(u);
          if (l) return a = setTimeout(z, t), b(u);
        }
        return void 0 === a && (a = setTimeout(z, t)), o;
      }
      return t = v(t) || 0, p(i) && (h = !!i.leading, s = (l = "maxWait" in i) ? f(v(i.maxWait) || 0, t) : s, y = "trailing" in i ? !!i.trailing : y), S.cancel = function () {
        void 0 !== a && clearTimeout(a), c = 0, n = u = r = a = void 0;
      }, S.flush = function () {
        return void 0 === a ? o : w(d());
      }, S;
    };
  }()),
  K = "sash-module_sash__K-9lB",
  Z = "sash-module_disabled__Hm-wx",
  Q = "sash-module_mac__Jf6OJ",
  q = "sash-module_vertical__pB-rs",
  ee = "sash-module_minimum__-UKxp",
  te = "sash-module_maximum__TCWxD",
  ie = "sash-module_horizontal__kFbiw",
  ne = "sash-module_hover__80W6I",
  re = "sash-module_active__bJspD";
var se = function (e) {
    return e.Vertical = "VERTICAL", e.Horizontal = "HORIZONTAL", e;
  }({}),
  oe = function (e) {
    return e.Disabled = "DISABLED", e.Minimum = "MINIMUM", e.Maximum = "MAXIMUM", e.Enabled = "ENABLED", e;
  }({}),
  ae = A ? 20 : 8;
var ue = new W();
class ce extends W {
  get state() {
    return this._state;
  }
  set state(e) {
    this._state !== e && (this.el.classList.toggle(Z, e === oe.Disabled), this.el.classList.toggle("sash-disabled", e === oe.Disabled), this.el.classList.toggle(ee, e === oe.Minimum), this.el.classList.toggle("sash-minimum", e === oe.Minimum), this.el.classList.toggle(te, e === oe.Maximum), this.el.classList.toggle("sash-maximum", e === oe.Maximum), this._state = e, this.emit("enablementChange", e));
  }
  constructor(e, t, i) {
    var _i$orientation;
    super(), this.el = void 0, this.layoutProvider = void 0, this.orientation = void 0, this.size = void 0, this.hoverDelay = 300, this.hoverDelayer = J(e => e.classList.add("sash-hover", ne), this.hoverDelay), this._state = oe.Enabled, this.onPointerStart = e => {
      var t = e.pageX,
        i = e.pageY,
        n = {
          startX: t,
          currentX: t,
          startY: i,
          currentY: i
        };
      this.el.classList.add("sash-active", re), this.emit("start", n), this.el.setPointerCapture(e.pointerId);
      var r = e => {
          e.preventDefault();
          var n = {
            startX: t,
            currentX: e.pageX,
            startY: i,
            currentY: e.pageY
          };
          this.emit("change", n);
        },
        s = e => {
          e.preventDefault(), this.el.classList.remove("sash-active", re), this.hoverDelayer.cancel(), this.emit("end"), this.el.releasePointerCapture(e.pointerId), window.removeEventListener("pointermove", r), window.removeEventListener("pointerup", s);
        };
      window.addEventListener("pointermove", r), window.addEventListener("pointerup", s);
    }, this.onPointerDoublePress = () => {
      this.emit("reset");
    }, this.onMouseEnter = () => {
      this.el.classList.contains(re) ? (this.hoverDelayer.cancel(), this.el.classList.add("sash-hover", ne)) : this.hoverDelayer(this.el);
    }, this.onMouseLeave = () => {
      this.hoverDelayer.cancel(), this.el.classList.remove("sash-hover", ne);
    }, this.el = document.createElement("div"), this.el.classList.add("sash", K), this.el.dataset.testid = "sash", e.append(this.el), T && this.el.classList.add("sash-mac", Q), this.el.addEventListener("pointerdown", this.onPointerStart), this.el.addEventListener("dblclick", this.onPointerDoublePress), this.el.addEventListener("mouseenter", this.onMouseEnter), this.el.addEventListener("mouseleave", this.onMouseLeave), "number" == typeof i.size ? (this.size = i.size, i.orientation === se.Vertical ? this.el.style.width = this.size + "px" : this.el.style.height = this.size + "px") : (this.size = ae, ue.on("onDidChangeGlobalSize", e => {
      this.size = e, this.layout();
    })), this.layoutProvider = t, this.orientation = (_i$orientation = i.orientation) != null ? _i$orientation : se.Vertical, this.orientation === se.Horizontal ? (this.el.classList.add("sash-horizontal", ie), this.el.classList.remove("sash-vertical", q)) : (this.el.classList.remove("sash-horizontal", ie), this.el.classList.add("sash-vertical", q)), this.layout();
  }
  layout() {
    if (this.orientation === se.Vertical) {
      var _e2 = this.layoutProvider;
      this.el.style.left = _e2.getVerticalSashLeft(this) - this.size / 2 + "px", _e2.getVerticalSashTop && (this.el.style.top = _e2.getVerticalSashTop(this) + "px"), _e2.getVerticalSashHeight && (this.el.style.height = _e2.getVerticalSashHeight(this) + "px");
    } else {
      var _e3 = this.layoutProvider;
      this.el.style.top = _e3.getHorizontalSashTop(this) - this.size / 2 + "px", _e3.getHorizontalSashLeft && (this.el.style.left = _e3.getHorizontalSashLeft(this) + "px"), _e3.getHorizontalSashWidth && (this.el.style.width = _e3.getHorizontalSashWidth(this) + "px");
    }
  }
  dispose() {
    this.el.removeEventListener("pointerdown", this.onPointerStart), this.el.removeEventListener("dblclick", this.onPointerDoublePress), this.el.removeEventListener("mouseenter", this.onMouseEnter), this.el.removeEventListener("mouseleave", () => this.onMouseLeave), this.el.remove();
  }
}
var he;
var le;
(le = he || (he = {})).Distribute = {
  type: "distribute"
}, le.Split = function (e) {
  return {
    type: "split",
    index: e
  };
}, le.Invisible = function (e) {
  return {
    type: "invisible",
    cachedVisibleSize: e
  };
};
var fe = function (e) {
  return e.Normal = "NORMAL", e.Low = "LOW", e.High = "HIGH", e;
}({});
class me {
  constructor(e, t, i) {
    this.container = void 0, this.view = void 0, this._size = void 0, this._cachedVisibleSize = void 0, this.container = e, this.view = t, this.container.classList.add("split-view-view", j), this.container.dataset.testid = "split-view-view", "number" == typeof i ? (this._size = i, this._cachedVisibleSize = void 0, e.classList.add("split-view-view-visible")) : (this._size = 0, this._cachedVisibleSize = i.cachedVisibleSize);
  }
  set size(e) {
    this._size = e;
  }
  get size() {
    return this._size;
  }
  get priority() {
    return this.view.priority;
  }
  get snap() {
    return !!this.view.snap;
  }
  get cachedVisibleSize() {
    return this._cachedVisibleSize;
  }
  get visible() {
    return void 0 === this._cachedVisibleSize;
  }
  setVisible(e, t) {
    e !== this.visible && (e ? (this.size = g(this._cachedVisibleSize, this.viewMinimumSize, this.viewMaximumSize), this._cachedVisibleSize = void 0) : (this._cachedVisibleSize = "number" == typeof t ? t : this.size, this.size = 0), this.container.classList.toggle("split-view-view-visible", e), this.view.setVisible && this.view.setVisible(e));
  }
  get minimumSize() {
    return this.visible ? this.view.minimumSize : 0;
  }
  get viewMinimumSize() {
    return this.view.minimumSize;
  }
  get maximumSize() {
    return this.visible ? this.view.maximumSize : 0;
  }
  get viewMaximumSize() {
    return this.view.maximumSize;
  }
  set enabled(e) {
    this.container.style.pointerEvents = e ? "" : "none";
  }
  layout(e) {
    this.layoutContainer(e), this.view.layout(this.size, e);
  }
}
class de extends me {
  layoutContainer(e) {
    this.container.style.left = e + "px", this.container.style.width = this.size + "px";
  }
}
class pe extends me {
  layoutContainer(e) {
    this.container.style.top = e + "px", this.container.style.height = this.size + "px";
  }
}
class ve extends W {
  get startSnappingEnabled() {
    return this._startSnappingEnabled;
  }
  set startSnappingEnabled(e) {
    this._startSnappingEnabled !== e && (this._startSnappingEnabled = e, this.updateSashEnablement());
  }
  get endSnappingEnabled() {
    return this._endSnappingEnabled;
  }
  set endSnappingEnabled(e) {
    this._endSnappingEnabled !== e && (this._endSnappingEnabled = e, this.updateSashEnablement());
  }
  constructor(e, t, i, n, r) {
    var _t$orientation, _t$proportionalLayout;
    if (t === void 0) {
      t = {};
    }
    if (super(), this.onDidChange = void 0, this.onDidDragStart = void 0, this.onDidDragEnd = void 0, this.orientation = void 0, this.sashContainer = void 0, this.size = 0, this.contentSize = 0, this.proportions = void 0, this.viewItems = [], this.sashItems = [], this.sashDragState = void 0, this.proportionalLayout = void 0, this.getSashOrthogonalSize = void 0, this._startSnappingEnabled = !0, this._endSnappingEnabled = !0, this.onSashEnd = e => {
      this.emit("sashchange", e), this.saveProportions();
      for (var _e4 of this.viewItems) _e4.enabled = !0;
    }, this.orientation = (_t$orientation = t.orientation) != null ? _t$orientation : se.Vertical, this.proportionalLayout = (_t$proportionalLayout = t.proportionalLayout) != null ? _t$proportionalLayout : !0, this.getSashOrthogonalSize = t.getSashOrthogonalSize, i && (this.onDidChange = i), n && (this.onDidDragStart = n), r && (this.onDidDragEnd = r), this.sashContainer = document.createElement("div"), this.sashContainer.classList.add("sash-container", E), e.prepend(this.sashContainer), t.descriptor) {
      this.size = t.descriptor.size;
      for (var [_e5, _i] of t.descriptor.views.entries()) {
        var _t = _i.size,
          _n = _i.container,
          _r = _i.view;
        this.addView(_n, _r, _t, _e5, !0);
      }
      this.contentSize = this.viewItems.reduce((e, t) => e + t.size, 0), this.saveProportions();
    }
  }
  addView(e, t, i, n, r) {
    if (n === void 0) {
      n = this.viewItems.length;
    }
    var s;
    s = "number" == typeof i ? i : "split" === i.type ? this.getViewSize(i.index) / 2 : "invisible" === i.type ? {
      cachedVisibleSize: i.cachedVisibleSize
    } : t.minimumSize;
    var o = this.orientation === se.Vertical ? new pe(e, t, s) : new de(e, t, s);
    if (this.viewItems.splice(n, 0, o), this.viewItems.length > 1) {
      var _e6 = this.orientation === se.Vertical ? new ce(this.sashContainer, {
          getHorizontalSashTop: e => this.getSashPosition(e),
          getHorizontalSashWidth: this.getSashOrthogonalSize
        }, {
          orientation: se.Horizontal
        }) : new ce(this.sashContainer, {
          getVerticalSashLeft: e => this.getSashPosition(e),
          getVerticalSashHeight: this.getSashOrthogonalSize
        }, {
          orientation: se.Vertical
        }),
        _t2 = this.orientation === se.Vertical ? t => ({
          sash: _e6,
          start: t.startY,
          current: t.currentY
        }) : t => ({
          sash: _e6,
          start: t.startX,
          current: t.currentX
        });
      _e6.on("start", e => {
        var _this$onDidDragStart;
        this.emit("sashDragStart"), this.onSashStart(_t2(e));
        var i = this.viewItems.map(e => e.size);
        (_this$onDidDragStart = this.onDidDragStart) == null || _this$onDidDragStart.call(this, i);
      }), _e6.on("change", e => this.onSashChange(_t2(e))), _e6.on("end", () => {
        var _this$onDidDragEnd;
        this.emit("sashDragEnd"), this.onSashEnd(this.sashItems.findIndex(t => t.sash === _e6));
        var t = this.viewItems.map(e => e.size);
        (_this$onDidDragEnd = this.onDidDragEnd) == null || _this$onDidDragEnd.call(this, t);
      }), _e6.on("reset", () => {
        var t = this.sashItems.findIndex(t => t.sash === _e6),
          i = X(t, -1, -1),
          n = X(t + 1, this.viewItems.length),
          r = this.findFirstSnapIndex(i),
          s = this.findFirstSnapIndex(n);
        ("number" != typeof r || this.viewItems[r].visible) && ("number" != typeof s || this.viewItems[s].visible) && this.emit("sashreset", t);
      });
      var _i2 = {
        sash: _e6
      };
      this.sashItems.splice(n - 1, 0, _i2);
    }
    r || this.relayout(), r || "number" == typeof i || "distribute" !== i.type || this.distributeViewSizes();
  }
  removeView(e, t) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    var i = this.viewItems.splice(e, 1)[0].view;
    if (this.viewItems.length >= 1) {
      var _t3 = Math.max(e - 1, 0);
      this.sashItems.splice(_t3, 1)[0].sash.dispose();
    }
    return this.relayout(), t && "distribute" === t.type && this.distributeViewSizes(), i;
  }
  moveView(e, t, i) {
    var n = this.getViewCachedVisibleSize(t),
      r = void 0 === n ? this.getViewSize(t) : he.Invisible(n),
      s = this.removeView(t);
    this.addView(e, s, r, i);
  }
  getViewCachedVisibleSize(e) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    return this.viewItems[e].cachedVisibleSize;
  }
  layout(e) {
    if (e === void 0) {
      e = this.size;
    }
    var t = Math.max(this.size, this.contentSize);
    if (this.size = e, this.proportions) for (var _t4 = 0; _t4 < this.viewItems.length; _t4++) {
      var _i3 = this.viewItems[_t4];
      _i3.size = g(Math.round(this.proportions[_t4] * e), _i3.minimumSize, _i3.maximumSize);
    } else {
      var _i4 = X(0, this.viewItems.length),
        _n2 = _i4.filter(e => this.viewItems[e].priority === fe.Low),
        _r2 = _i4.filter(e => this.viewItems[e].priority === fe.High);
      this.resize(this.viewItems.length - 1, e - t, void 0, _n2, _r2);
    }
    this.distributeEmptySpace(), this.layoutViews();
  }
  resizeView(e, t) {
    if (e < 0 || e >= this.viewItems.length) return;
    var i = X(0, this.viewItems.length).filter(t => t !== e),
      n = [...i.filter(e => this.viewItems[e].priority === fe.Low), e],
      r = i.filter(e => this.viewItems[e].priority === fe.High),
      s = this.viewItems[e];
    t = Math.round(t), t = g(t, s.minimumSize, Math.min(s.maximumSize, this.size)), s.size = t, this.relayout(n, r);
  }
  resizeViews(e) {
    for (var _t5 = 0; _t5 < e.length; _t5++) {
      var _i5 = this.viewItems[_t5];
      var _n3 = e[_t5];
      _n3 = Math.round(_n3), _n3 = g(_n3, _i5.minimumSize, Math.min(_i5.maximumSize, this.size)), _i5.size = _n3;
    }
    this.contentSize = this.viewItems.reduce((e, t) => e + t.size, 0), this.saveProportions(), this.layout(this.size);
  }
  getViewSize(e) {
    return e < 0 || e >= this.viewItems.length ? -1 : this.viewItems[e].size;
  }
  isViewVisible(e) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    return this.viewItems[e].visible;
  }
  setViewVisible(e, t) {
    if (e < 0 || e >= this.viewItems.length) throw new Error("Index out of bounds");
    this.viewItems[e].setVisible(t), this.distributeEmptySpace(e), this.layoutViews(), this.saveProportions();
  }
  distributeViewSizes() {
    var e = [];
    var t = 0;
    for (var _i6 of this.viewItems) _i6.maximumSize - _i6.minimumSize > 0 && (e.push(_i6), t += _i6.size);
    var i = Math.floor(t / e.length);
    for (var _t6 of e) _t6.size = g(i, _t6.minimumSize, _t6.maximumSize);
    var n = X(0, this.viewItems.length),
      r = n.filter(e => this.viewItems[e].priority === fe.Low),
      s = n.filter(e => this.viewItems[e].priority === fe.High);
    this.relayout(r, s);
  }
  dispose() {
    this.sashItems.forEach(e => e.sash.dispose()), this.sashItems = [], this.sashContainer.remove();
  }
  relayout(e, t) {
    var i = this.viewItems.reduce((e, t) => e + t.size, 0);
    this.resize(this.viewItems.length - 1, this.size - i, void 0, e, t), this.distributeEmptySpace(), this.layoutViews(), this.saveProportions();
  }
  onSashStart(_ref) {
    var {
      sash: e,
      start: t
    } = _ref;
    var i = this.sashItems.findIndex(t => t.sash === e);
    (e => {
      var t = this.viewItems.map(e => e.size);
      var n,
        r,
        s = Number.NEGATIVE_INFINITY,
        o = Number.POSITIVE_INFINITY;
      var a = X(i, -1, -1),
        u = X(i + 1, this.viewItems.length),
        c = a.reduce((e, i) => e + (this.viewItems[i].minimumSize - t[i]), 0),
        h = a.reduce((e, i) => e + (this.viewItems[i].viewMaximumSize - t[i]), 0),
        l = 0 === u.length ? Number.POSITIVE_INFINITY : u.reduce((e, i) => e + (t[i] - this.viewItems[i].minimumSize), 0),
        f = 0 === u.length ? Number.NEGATIVE_INFINITY : u.reduce((e, i) => e + (t[i] - this.viewItems[i].viewMaximumSize), 0);
      s = Math.max(c, f), o = Math.min(l, h);
      var m = this.findFirstSnapIndex(a),
        d = this.findFirstSnapIndex(u);
      if ("number" == typeof m) {
        var _e7 = this.viewItems[m],
          _t7 = Math.floor(_e7.viewMinimumSize / 2);
        n = {
          index: m,
          limitDelta: _e7.visible ? s - _t7 : s + _t7,
          size: _e7.size
        };
      }
      if ("number" == typeof d) {
        var _e8 = this.viewItems[d],
          _t8 = Math.floor(_e8.viewMinimumSize / 2);
        r = {
          index: d,
          limitDelta: _e8.visible ? o + _t8 : o - _t8,
          size: _e8.size
        };
      }
      this.sashDragState = {
        start: e,
        current: e,
        index: i,
        sizes: t,
        minDelta: s,
        maxDelta: o,
        snapBefore: n,
        snapAfter: r
      };
    })(t);
  }
  onSashChange(_ref2) {
    var {
      current: e
    } = _ref2;
    var {
      index: t,
      start: i,
      sizes: n,
      minDelta: r,
      maxDelta: s,
      snapBefore: o,
      snapAfter: a
    } = this.sashDragState;
    this.sashDragState.current = e;
    var u = e - i;
    this.resize(t, u, n, void 0, void 0, r, s, o, a), this.distributeEmptySpace(), this.layoutViews();
  }
  getSashPosition(e) {
    var t = 0;
    for (var _i7 = 0; _i7 < this.sashItems.length; _i7++) if (t += this.viewItems[_i7].size, this.sashItems[_i7].sash === e) return t;
    return 0;
  }
  resize(e, t, i, n, r, s, o, a, u) {
    if (i === void 0) {
      i = this.viewItems.map(e => e.size);
    }
    if (s === void 0) {
      s = Number.NEGATIVE_INFINITY;
    }
    if (o === void 0) {
      o = Number.POSITIVE_INFINITY;
    }
    if (e < 0 || e >= this.viewItems.length) return 0;
    var c = X(e, -1, -1),
      h = X(e + 1, this.viewItems.length);
    if (r) for (var _e9 of r) G(c, _e9), G(h, _e9);
    if (n) for (var _e10 of n) U(c, _e10), U(h, _e10);
    var l = c.map(e => this.viewItems[e]),
      f = c.map(e => i[e]),
      m = h.map(e => this.viewItems[e]),
      d = h.map(e => i[e]),
      p = c.reduce((e, t) => e + (this.viewItems[t].minimumSize - i[t]), 0),
      v = c.reduce((e, t) => e + (this.viewItems[t].maximumSize - i[t]), 0),
      y = 0 === h.length ? Number.POSITIVE_INFINITY : h.reduce((e, t) => e + (i[t] - this.viewItems[t].minimumSize), 0),
      b = 0 === h.length ? Number.NEGATIVE_INFINITY : h.reduce((e, t) => e + (i[t] - this.viewItems[t].maximumSize), 0),
      z = Math.max(p, b, s),
      w = Math.min(y, v, o);
    var S = !1;
    if (a) {
      var _e11 = this.viewItems[a.index],
        _i8 = t >= a.limitDelta;
      S = _i8 !== _e11.visible, _e11.setVisible(_i8, a.size);
    }
    if (!S && u) {
      var _e12 = this.viewItems[u.index],
        _i9 = t < u.limitDelta;
      S = _i9 !== _e12.visible, _e12.setVisible(_i9, u.size);
    }
    if (S) return this.resize(e, t, i, n, r, s, o);
    for (var _e13 = 0, _i10 = t = g(t, z, w); _e13 < l.length; _e13++) {
      var _t9 = l[_e13],
        _n4 = g(f[_e13] + _i10, _t9.minimumSize, _t9.maximumSize);
      _i10 -= _n4 - f[_e13], _t9.size = _n4;
    }
    for (var _e14 = 0, _i11 = t; _e14 < m.length; _e14++) {
      var _t10 = m[_e14],
        _n5 = g(d[_e14] - _i11, _t10.minimumSize, _t10.maximumSize);
      _i11 += _n5 - d[_e14], _t10.size = _n5;
    }
    return t;
  }
  distributeEmptySpace(e) {
    var t = this.viewItems.reduce((e, t) => e + t.size, 0);
    var i = this.size - t;
    var n = X(this.viewItems.length - 1, -1, -1);
    "number" == typeof e && U(n, e);
    for (var _e15 = 0; 0 !== i && _e15 < n.length; _e15++) {
      var _t11 = this.viewItems[n[_e15]],
        _r3 = g(_t11.size + i, _t11.minimumSize, _t11.maximumSize);
      i -= _r3 - _t11.size, _t11.size = _r3;
    }
  }
  layoutViews() {
    var _this$onDidChange;
    this.contentSize = this.viewItems.reduce((e, t) => e + t.size, 0);
    var e = 0;
    for (var _t12 of this.viewItems) _t12.layout(e), e += _t12.size;
    (_this$onDidChange = this.onDidChange) != null && _this$onDidChange.call(this, this.viewItems.map(e => e.size)), this.sashItems.forEach(e => e.sash.layout()), this.updateSashEnablement();
  }
  saveProportions() {
    this.proportionalLayout && this.contentSize > 0 && (this.proportions = this.viewItems.map(e => e.size / this.contentSize));
  }
  updateSashEnablement() {
    var e = !1;
    var t = this.viewItems.map(t => e = t.size - t.minimumSize > 0 || e);
    e = !1;
    var i = this.viewItems.map(t => e = t.maximumSize - t.size > 0 || e),
      n = [...this.viewItems].reverse();
    e = !1;
    var r = n.map(t => e = t.size - t.minimumSize > 0 || e).reverse();
    e = !1;
    var s = n.map(t => e = t.maximumSize - t.size > 0 || e).reverse();
    var o = 0;
    for (var _e16 = 0; _e16 < this.sashItems.length; _e16++) {
      var {
        sash: _n6
      } = this.sashItems[_e16];
      o += this.viewItems[_e16].size;
      var _a = !(t[_e16] && s[_e16 + 1]),
        _u = !(i[_e16] && r[_e16 + 1]);
      if (_a && _u) {
        var _i12 = X(_e16, -1, -1),
          _s = X(_e16 + 1, this.viewItems.length),
          _a2 = this.findFirstSnapIndex(_i12),
          _u2 = this.findFirstSnapIndex(_s),
          _c = "number" == typeof _a2 && !this.viewItems[_a2].visible,
          _h = "number" == typeof _u2 && !this.viewItems[_u2].visible;
        _c && r[_e16] && (o > 0 || this.startSnappingEnabled) ? _n6.state = oe.Minimum : _h && t[_e16] && (o < this.contentSize || this.endSnappingEnabled) ? _n6.state = oe.Maximum : _n6.state = oe.Disabled;
      } else _n6.state = _a && !_u ? oe.Minimum : !_a && _u ? oe.Maximum : oe.Enabled;
    }
  }
  findFirstSnapIndex(e) {
    for (var _t13 of e) {
      var _e17 = this.viewItems[_t13];
      if (_e17.visible && _e17.snap) return _t13;
    }
    for (var _t14 of e) {
      var _e18 = this.viewItems[_t14];
      if (_e18.visible && _e18.maximumSize - _e18.minimumSize > 0) return;
      if (!_e18.visible && _e18.snap) return _t14;
    }
  }
}
class ye {
  constructor(e) {
    this.size = void 0, this.size = e;
  }
  getPreferredSize() {
    return this.size;
  }
}
class be {
  constructor(e, t) {
    this.proportion = void 0, this.layoutService = void 0, this.proportion = e, this.layoutService = t;
  }
  getPreferredSize() {
    return this.proportion * this.layoutService.getSize();
  }
}
class ge {
  getPreferredSize() {}
}
class ze {
  get preferredSize() {
    return this.layoutStrategy.getPreferredSize();
  }
  set preferredSize(e) {
    if ("number" == typeof e) this.layoutStrategy = new ye(e);else if ("string" == typeof e) {
      var _t15 = e.trim();
      if (k(_t15, "%")) {
        var _e19 = Number(_t15.slice(0, -1)) / 100;
        this.layoutStrategy = new be(_e19, this.layoutService);
      } else if (k(_t15, "px")) {
        var _e20 = Number(_t15.slice(0, -2)) / 100;
        this.layoutStrategy = new ye(_e20);
      } else if ("number" == typeof Number.parseFloat(_t15)) {
        var _e21 = Number.parseFloat(_t15);
        this.layoutStrategy = new ye(_e21);
      } else this.layoutStrategy = new ge();
    } else this.layoutStrategy = new ge();
  }
  constructor(e, t) {
    var _t$priority;
    if (this.minimumSize = 0, this.maximumSize = Number.POSITIVE_INFINITY, this.element = void 0, this.priority = void 0, this.snap = void 0, this.layoutService = void 0, this.layoutStrategy = void 0, this.layoutService = e, this.element = t.element, this.minimumSize = "number" == typeof t.minimumSize ? t.minimumSize : 30, this.maximumSize = "number" == typeof t.maximumSize ? t.maximumSize : Number.POSITIVE_INFINITY, "number" == typeof t.preferredSize) this.layoutStrategy = new ye(t.preferredSize);else if ("string" == typeof t.preferredSize) {
      var _e22 = t.preferredSize.trim();
      if (k(_e22, "%")) {
        var _t16 = Number(_e22.slice(0, -1)) / 100;
        this.layoutStrategy = new be(_t16, this.layoutService);
      } else if (k(_e22, "px")) {
        var _t17 = Number(_e22.slice(0, -2));
        this.layoutStrategy = new ye(_t17);
      } else if ("number" == typeof Number.parseFloat(_e22)) {
        var _t18 = Number.parseFloat(_e22);
        this.layoutStrategy = new ye(_t18);
      } else this.layoutStrategy = new ge();
    } else this.layoutStrategy = new ge();
    this.priority = (_t$priority = t.priority) != null ? _t$priority : fe.Normal, this.snap = "boolean" == typeof t.snap && t.snap;
  }
  layout(e) {}
}
function we(e) {
  return void 0 !== e.minSize || void 0 !== e.maxSize || void 0 !== e.preferredSize || void 0 !== e.priority || void 0 !== e.visible;
}
var Se = a((_ref3, n) => {
  var {
    className: t,
    children: i
  } = _ref3;
  return e.createElement("div", {
    ref: n,
    className: y("split-view-view", j, t)
  }, i);
});
Se.displayName = "Allotment.Pane";
var _e = a((_ref4, j) => {
  var {
    children: o,
    className: a,
    id: c,
    maxSize: h = 1 / 0,
    minSize: l = 30,
    proportionalLayout: f = !0,
    separator: m = !0,
    sizes: d,
    defaultSizes: p = d,
    snap: v = !1,
    vertical: b = !1,
    onChange: g,
    onReset: z,
    onVisibleChange: w,
    onDragStart: _,
    onDragEnd: E
  } = _ref4;
  var N = t(null),
    M = t([]),
    P = t(new Map()),
    T = t(null),
    k = t(new Map()),
    $ = t(new F()),
    H = t([]),
    [Y, B] = i(!1);
  "production" !== process.env.NODE_ENV && d && console.warn("Prop sizes is deprecated. Please use defaultSizes instead.");
  var R = s(() => e.Children.toArray(o).filter(e.isValidElement), [o]),
    W = r(e => {
      var _H$current, _T$current;
      var t = (_H$current = H.current) == null ? void 0 : _H$current[e];
      return "number" == typeof (t == null ? void 0 : t.preferredSize) && ((_T$current = T.current) != null && _T$current.resizeView(e, Math.round(t.preferredSize)), !0);
    }, []);
  return u(j, () => ({
    reset: () => {
      if (z) z();else {
        var _T$current2;
        (_T$current2 = T.current) == null || _T$current2.distributeViewSizes();
        for (var _e23 = 0; _e23 < H.current.length; _e23++) W(_e23);
      }
    },
    resize: e => {
      var _T$current3;
      (_T$current3 = T.current) == null || _T$current3.resizeViews(e);
    }
  })), C(() => {
    var e = !0;
    p && k.current.size !== p.length && (e = !1, console.warn("Expected " + p.length + " children based on defaultSizes but found " + k.current.size)), e && p && (M.current = R.map(e => e.key));
    var t = _extends({
      orientation: b ? se.Vertical : se.Horizontal,
      proportionalLayout: f
    }, e && p && {
      descriptor: {
        size: p.reduce((e, t) => e + t, 0),
        views: p.map((e, t) => {
          var _i$minSize, _i$maxSize, _i$priority, _i$snap;
          var i = P.current.get(M.current[t]),
            n = new ze($.current, _extends({
              element: document.createElement("div"),
              minimumSize: (_i$minSize = i == null ? void 0 : i.minSize) != null ? _i$minSize : l,
              maximumSize: (_i$maxSize = i == null ? void 0 : i.maxSize) != null ? _i$maxSize : h,
              priority: (_i$priority = i == null ? void 0 : i.priority) != null ? _i$priority : fe.Normal
            }, (i == null ? void 0 : i.preferredSize) && {
              preferredSize: i == null ? void 0 : i.preferredSize
            }, {
              snap: (_i$snap = i == null ? void 0 : i.snap) != null ? _i$snap : v
            }));
          return H.current.push(n), {
            container: [...k.current.values()][t],
            size: e,
            view: n
          };
        })
      }
    });
    T.current = new ve(N.current, t, g, _, E), T.current.on("sashDragStart", () => {
      var _N$current;
      (_N$current = N.current) == null || _N$current.classList.add("split-view-sash-dragging");
    }), T.current.on("sashDragEnd", () => {
      var _N$current2;
      (_N$current2 = N.current) == null || _N$current2.classList.remove("split-view-sash-dragging");
    }), T.current.on("sashchange", e => {
      if (w && T.current) {
        var _e24 = R.map(e => e.key);
        for (var _t19 = 0; _t19 < _e24.length; _t19++) {
          var _i13 = P.current.get(_e24[_t19]);
          void 0 !== (_i13 == null ? void 0 : _i13.visible) && _i13.visible !== T.current.isViewVisible(_t19) && w(_t19, T.current.isViewVisible(_t19));
        }
      }
    }), T.current.on("sashreset", e => {
      if (z) z();else {
        var _T$current4;
        if (W(e)) return;
        if (W(e + 1)) return;
        (_T$current4 = T.current) == null || _T$current4.distributeViewSizes();
      }
    });
    var i = T.current;
    return () => {
      i.dispose();
    };
  }, []), C(() => {
    if (Y) {
      var _e25 = R.map(e => e.key),
        _t20 = [...M.current],
        _i14 = _e25.filter(e => !M.current.includes(e)),
        _n7 = _e25.filter(e => M.current.includes(e)),
        _r4 = M.current.map(t => !_e25.includes(t));
      for (var _e26 = _r4.length - 1; _e26 >= 0; _e26--) {
        var _T$current5;
        _r4[_e26] && ((_T$current5 = T.current) != null && _T$current5.removeView(_e26), _t20.splice(_e26, 1), H.current.splice(_e26, 1));
      }
      var _loop = function _loop(_n8) {
        var _i$minSize2, _i$maxSize2, _i$priority2, _i$snap2, _T$current7;
        var i = P.current.get(_n8),
          r = new ze($.current, _extends({
            element: document.createElement("div"),
            minimumSize: (_i$minSize2 = i == null ? void 0 : i.minSize) != null ? _i$minSize2 : l,
            maximumSize: (_i$maxSize2 = i == null ? void 0 : i.maxSize) != null ? _i$maxSize2 : h,
            priority: (_i$priority2 = i == null ? void 0 : i.priority) != null ? _i$priority2 : fe.Normal
          }, (i == null ? void 0 : i.preferredSize) && {
            preferredSize: i == null ? void 0 : i.preferredSize
          }, {
            snap: (_i$snap2 = i == null ? void 0 : i.snap) != null ? _i$snap2 : v
          }));
        (_T$current7 = T.current) != null && _T$current7.addView(k.current.get(_n8), r, he.Distribute, _e25.findIndex(e => e === _n8)), _t20.splice(_e25.findIndex(e => e === _n8), 0, _n8), H.current.splice(_e25.findIndex(e => e === _n8), 0, r);
      };
      for (var _n8 of _i14) {
        _loop(_n8);
      }
      for (; !S(_e25, _t20);) {
        var _loop2 = function _loop2(_n9) {
          var e = _t20.findIndex(e => e === _n9);
          if (e !== _i15) {
            var _T$current6;
            (_T$current6 = T.current) == null || _T$current6.moveView(k.current.get(_n9), e, _i15);
            var _r5 = _t20[e];
            _t20.splice(e, 1), _t20.splice(_i15, 0, _r5);
            return 1; // break
          }
        };
        for (var [_i15, _n9] of _e25.entries()) {
          if (_loop2(_n9)) break;
        }
      }
      var _loop3 = function _loop3(_t21) {
        var _T$current8;
        var i = _e25.findIndex(e => e === _t21),
          n = H.current[i].preferredSize;
        void 0 !== n && ((_T$current8 = T.current) == null ? void 0 : _T$current8.resizeView(i, n));
      };
      for (var _t21 of _i14) {
        _loop3(_t21);
      }
      var _loop4 = function _loop4(_t22) {
        var _T$current9, _T$current10;
        var i = P.current.get(_t22),
          n = _e25.findIndex(e => e === _t22);
        i && we(i) && void 0 !== i.visible && ((_T$current9 = T.current) == null ? void 0 : _T$current9.isViewVisible(n)) !== i.visible && ((_T$current10 = T.current) == null ? void 0 : _T$current10.setViewVisible(n, i.visible));
      };
      for (var _t22 of [..._i14, ..._n7]) {
        _loop4(_t22);
      }
      var _loop5 = function _loop5(_t23) {
        var i = P.current.get(_t23),
          n = _e25.findIndex(e => e === _t23);
        if (i && we(i)) {
          var _T$current11;
          void 0 !== i.preferredSize && H.current[n].preferredSize !== i.preferredSize && (H.current[n].preferredSize = i.preferredSize);
          var _e27 = !1;
          void 0 !== i.minSize && H.current[n].minimumSize !== i.minSize && (H.current[n].minimumSize = i.minSize, _e27 = !0), void 0 !== i.maxSize && H.current[n].maximumSize !== i.maxSize && (H.current[n].maximumSize = i.maxSize, _e27 = !0), _e27 && ((_T$current11 = T.current) == null ? void 0 : _T$current11.layout());
        }
      };
      for (var _t23 of _n7) {
        _loop5(_t23);
      }
      (_i14.length > 0 || _r4.length > 0) && (M.current = _e25);
    }
  }, [R, Y, h, l, v]), n(() => {
    T.current && (T.current.onDidChange = g);
  }, [g]), n(() => {
    T.current && (T.current.onDidDragStart = _);
  }, [_]), n(() => {
    T.current && (T.current.onDidDragEnd = E);
  }, [E]), I({
    ref: N,
    onResize: _ref5 => {
      var _T$current12;
      var {
        width: e,
        height: t
      } = _ref5;
      e && t && ((_T$current12 = T.current) != null && _T$current12.layout(b ? t : e), $.current.setSize(b ? t : e), B(!0));
    }
  }), n(() => {
    A && Ie(20);
  }, []), e.createElement("div", {
    ref: N,
    className: y("split-view", b ? "split-view-vertical" : "split-view-horizontal", {
      "split-view-separator-border": m
    }, x, b ? L : O, {
      [D]: m
    }, a),
    id: c
  }, e.createElement("div", {
    className: y("split-view-container", V)
  }, e.Children.toArray(o).map(t => {
    if (!e.isValidElement(t)) return null;
    var i = t.key;
    return "Allotment.Pane" === t.type.displayName ? (P.current.set(i, t.props), e.cloneElement(t, {
      key: i,
      ref: e => {
        var n = t.ref;
        n && (n.current = e), e ? k.current.set(i, e) : k.current.delete(i);
      }
    })) : e.createElement(Se, {
      key: i,
      ref: e => {
        e ? k.current.set(i, e) : k.current.delete(i);
      }
    }, t);
  })));
});
function Ie(e) {
  var t = g(e, 4, 20),
    i = g(e, 1, 8);
  document.documentElement.style.setProperty("--sash-size", t + "px"), document.documentElement.style.setProperty("--sash-hover-size", i + "px"), function (e) {
    ae = e, ue.emit("onDidChangeGlobalSize", e);
  }(t);
}
_e.displayName = "Allotment";
var xe = Object.assign(_e, {
  Pane: Se
});
export { xe as Allotment, fe as LayoutPriority, Ie as setSashSize };
